// flux-utils/dist/capacitor.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

// Dependencies

var _events = require('events');

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

// Class Variables
var CHANGE_EVENT = 'change';

/*
 * Flux Capacitors
 * // todo: write useful comments
 */
function createCapacitor(options, Dispatcher) {
    // validation
    if (!Dispatcher) throw new Error('You must provide a Flux Dispatcher to register this store with.');
    if (!_lodash2['default'].has(options, 'actionHandlers')) throw new Error('Your store must provide an `actionHandlers` property mapping handlers to flux `action.type` constants');

    //
    var emitter = new _events.EventEmitter();
    emitter.setMaxListeners(0);

    //
    var eventMethods = {
        emitChange: function emitChange() {
            emitter.emit(CHANGE_EVENT);
        },
        addChangeListener: function addChangeListener(callback) {
            emitter.on(CHANGE_EVENT, callback);
        },
        removeChangeListener: function removeChangeListener(callback) {
            emitter.removeListener(CHANGE_EVENT, callback);
        }
    };

    //
    var actionHandlers = options.actionHandlers;

    var storeOptions = _objectWithoutProperties(options, ['actionHandlers']);

    var store = _objectAssign2['default']({}, storeOptions, eventMethods);

    store.dispatchToken = Dispatcher.register(function (action) {
        if (_lodash2['default'].has(actionHandlers, action.type)) {
            var actionHandler = actionHandlers[action.type];
            actionHandler.call(store, action);
        }
    });

    return store;
}

/*
 * Flux Capacitors Mixin
 * // todo: write useful comments
 */
var capacitorMixin = {
    getInitialState: function getInitialState() {
        return this.getState();
    },
    componentDidMount: function componentDidMount() {
        var _this = this;

        if (!this.listenTo) {
            return;
        }
        if (_lodash2['default'].isArray(this.listenTo)) {
            _lodash2['default'].each(this.listenTo, function (store) {
                store.addChangeListener(_this.onChange);
            });
        } else {
            this.listenTo.addChangeListener(this.onChange);
        }
    },
    componentWillUnmount: function componentWillUnmount() {
        var _this2 = this;

        if (!this.listenTo) {
            return;
        }

        if (_lodash2['default'].isArray(this.listenTo)) {
            _lodash2['default'].each(this.listenTo, function (store) {
                store.removeChangeListener(_this2.onChange);
            });
        } else {
            this.listenTo.removeChangeListener(this.onChange);
        }
    },
    onChange: function onChange() {
        this.setState(this.getState());
    }
};

exports.Capacitor = createCapacitor;
exports.Mixin = capacitorMixin;
